Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,322 @@
+                Gauge(label: self.labels.label(for: \.startTimeSeconds)).record(metrics.startTimeSeconds)
+                Gauge(label: self.labels.label(for: \.cpuSecondsTotal)).record(metrics.cpuSeconds)
+                Gauge(label: self.labels.label(for: \.maxFileDescriptors)).record(metrics.maxFileDescriptors)
+                Gauge(label: self.labels.label(for: \.openFileDescriptors)).record(metrics.openFileDescriptors)
+            })
+        }
+
+        internal func cancelSystemMetrics() {
+            self.task?.cancel()
+            self.task = nil
+        }
+
+        internal func poll() {
+            self.queue.asyncAfter(deadline: .now() + self.timeInterval) {
+                guard let task = self.task else { return }
+                task.perform()
+                self.poll()
+            }
+        }
+        
+        func makeCounter(label: String, dimensions: [(String, String)]) -> CounterHandler {
+            self.underlying.makeCounter(label: label, dimensions: dimensions)
+        }
+        
+        func makeRecorder(label: String, dimensions: [(String, String)], aggregate: Bool) -> RecorderHandler {
+            self.underlying.makeRecorder(label: label, dimensions: dimensions, aggregate: aggregate)
+        }
+        
+        func makeTimer(label: String, dimensions: [(String, String)]) -> TimerHandler {
+            self.underlying.makeTimer(label: label, dimensions: dimensions)
+        }
+        
+        func destroyCounter(_ handler: CounterHandler) {
+            self.underlying.destroyCounter(handler)
+        }
+        
+        func destroyRecorder(_ handler: RecorderHandler) {
+            self.underlying.destroyRecorder(handler)
+        }
+        
+        func destroyTimer(_ handler: TimerHandler) {
+            self.underlying.destroyTimer(handler)
+        }
+    }
+}
+
+public protocol SystemMetricsProvider {
+    static func readSystemMetrics() -> SystemMetrics.Data?
+}
+
+public enum SystemMetrics {
+    /// Options used to bootstrap `SystemMetricsHandler`
+    ///
+    /// Libraries are advised to extend `SystemMetricsOptions` with a static instance for ease of use.
+    public struct Options {
+        let interval: DispatchTimeInterval
+        let metricsType: SystemMetricsProvider.Type?
+        let labels: Labels
+        
+        /// Create new instance of `SystemMetricsOptions`
+        ///
+        /// - parameters:
+        ///     - pollInterval: The interval at which system metrics should be updated.
+        ///     - systemMetricsType: The type of system metrics to use. If none is provided this defaults to
+        ///                          `LinuxSystemMetrics` on Linux platforms and `NOOPSystemMetrics` on all other platforms.
+        ///     - labels: The labels to use for generated system metrics.
+        public init(pollInterval interval: DispatchTimeInterval = .seconds(2), metricsType: SystemMetricsProvider.Type? = nil, labels: Labels) {
```

Comment:
see comment above about changing  SystemMetricsProvider.Type to a functional generator